File: app/src/entourage/java/social/entourage/android/guide/filter/GuideFilterAdapter.java
package social.entourage.android.guide.filter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import social.entourage.android.R;
import social.entourage.android.guide.PoiRenderer;

/**
 * Created by mihaiionescu on 28/03/2017.
 */

public class GuideFilterAdapter extends BaseAdapter {

    // ----------------------------------
    // Attributes
    // ----------------------------------

    private List<GuideFilterItem> items = new ArrayList<>();

    private OnCheckedChangeListener onCheckedChangeListener = new OnCheckedChangeListener();

    // ----------------------------------
    // Implementation
    // ----------------------------------


    public GuideFilterAdapter() {
        GuideFilter guideFilter = GuideFilter.getInstance();
        for (int i = 1; i < PoiRenderer.CategoryType.values().length; i++) {
            PoiRenderer.CategoryType categoryType = PoiRenderer.CategoryType.values()[i];
            items.add(new GuideFilterItem(categoryType, guideFilter.valueForCategoryId(categoryType.getCategoryId())));
        }
    }

    @Override
    public int getCount() {
        // We need to ignore the category at index zero
        return items.size();
    }

    @Override
    public GuideFilterItem getItem(final int position) {
        return items.get(position);
    }

    @Override
    public long getItemId(final int position) {
        return 0;
    }

    @Override
    public View getView(final int position, View view, final ViewGroup viewGroup) {
        GuideFilterViewHolder viewHolder;
        if (view == null) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.layout_filter_item, viewGroup, false);
            viewHolder = new GuideFilterViewHolder(view, onCheckedChangeListener);
            view.setTag(viewHolder);
        } else {
            viewHolder = (GuideFilterViewHolder) view.getTag();
        }

        // Populate the view
        GuideFilterItem item = getItem(position);
        if (item != null && item.categoryType != null) {
            PoiRenderer.CategoryType categoryType = item.categoryType;

            viewHolder.mFilterName.setText(categoryType.getName());
            viewHolder.mFilterImage.setImageResource(categoryType.getResourceId());

            // set the switch
            viewHolder.mFilterSwitch.setChecked(item.isChecked);
            viewHolder.mFilterSwitch.setTag(position);

            // separator is visible unless last element
            viewHolder.mSeparatorView.setVisibility(position == getCount()-1 ? View.GONE : View.VISIBLE);

            // set the category id
            viewHolder.categoryId = categoryType.getCategoryId();
        }

        return view;
    }

    // ----------------------------------
    // View Holder
    // ----------------------------------

    public static class GuideFilterViewHolder {

        public TextView mFilterName;
        public ImageView mFilterImage;
        public Switch mFilterSwitch;
        public View mSeparatorView;

        public long categoryId = 0;

        public GuideFilterViewHolder(View v, OnCheckedChangeListener onCheckedChangeListener) {
            mFilterName = v.findViewById(R.id.filter_item_text);
            mFilterImage = v.findViewById(R.id.filter_item_image);
            mFilterSwitch = v.findViewById(R.id.filter_item_switch);
            mSeparatorView = v.findViewById(R.id.filter_item_separator);

            mFilterSwitch.setOnCheckedChangeListener(onCheckedChangeListener);
        }

    }

    // ----------------------------------
    // Filter Item
    // ----------------------------------

    public static class GuideFilterItem {

        public PoiRenderer.CategoryType categoryType;
        public boolean isChecked;

        public GuideFilterItem(PoiRenderer.CategoryType categoryType, boolean isChecked) {
            this.categoryType = categoryType;
            this.isChecked = isChecked;
        }

    }

    private class OnCheckedChangeListener implements CompoundButton.OnCheckedChangeListener {
        @Override
        public void onCheckedChanged(final CompoundButton compoundButton, final boolean isChecked) {
            // if no tag, exit
            if (compoundButton.getTag() == null) {
                return;
            }
            // get the position
            int position = (Integer) compoundButton.getTag();

            GuideFilterItem item = GuideFilterAdapter.this.getItem(position);
            if (item == null) {
                return;
            }

            item.isChecked = isChecked;
        }
    }
}

File: app/src/main/java/social/entourage/android/guide/filter/GuideFilterAdapter.kt
package social.entourage.android.guide.filter

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import kotlinx.android.synthetic.main.layout_filter_item.view.*
import social.entourage.android.R
import social.entourage.android.guide.filter.GuideFilter.Companion.instance
import social.entourage.android.guide.poi.PoiRenderer.CategoryType

/**
 * Created by mihaiionescu on 28/03/2017.
 */
class GuideFilterAdapter(var context: Context) : BaseAdapter() {
    // ----------------------------------
    // Attributes
    // ----------------------------------
    val items: MutableList<GuideFilterItem> = ArrayList()

    var isHelpOnly = false

    fun setHelpOnly() {
        isHelpOnly = true
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        // We need to ignore the category at index zero
        return items.size
    }

    override fun getItem(position: Int): GuideFilterItem {
        return items[position]
    }

    override fun getItemId(position: Int): Long {
        //??
        return 0
    }

    override fun getView(position: Int, view2: View?, viewGroup: ViewGroup): View {
        var view = view2
        val viewHolder: GuideFilterViewHolder
        if (view == null) {
            view = LayoutInflater.from(viewGroup.context).inflate(R.layout.layout_filter_item, viewGroup, false)
            viewHolder = GuideFilterViewHolder(view as View)
            view.tag = viewHolder
        }
        // Populate the view
        val item = getItem(position)
        val categoryDisplayName = when (item.categoryType.displayName) {
            "Other" -> context.getString(R.string.category_other)
            "Se nourrir" -> context.getString(R.string.category_food)
            "Se loger" -> context.getString(R.string.category_housing)
            "Se soigner" -> context.getString(R.string.category_medical)
            "S'orienter" -> context.getString(R.string.category_orientation)
            "Se réinsérer" -> context.getString(R.string.category_insertion)
            "Partenaires" -> context.getString(R.string.category_partners)
            "Toilettes" -> context.getString(R.string.category_toilettes)
            "Fontaines" -> context.getString(R.string.category_fontaines)
            "Douches" -> context.getString(R.string.category_douches)
            "Laveries" -> context.getString(R.string.category_laverlinge)
            "Bien-être & activités" -> context.getString(R.string.category_self_care)
            "Vêtements & matériels" -> context.getString(R.string.category_vetements)
            "Bagageries" -> context.getString(R.string.category_bagages)
            "Boîtes à dons & lire" -> context.getString(R.string.category_boitesdons)
            else -> item.categoryType.displayName
        }


        val categoryType = item.categoryType
        val displayName: String = if (categoryType.categoryId == CategoryType.PARTNERS.categoryId) {
            view.context?.getString(R.string.guide_display_partners) ?: categoryDisplayName
        } else categoryDisplayName
        view.filter_item_text?.text = displayName
        view.filter_item_image?.setImageResource(categoryType.filterId)
        // set the switch
        view.filter_item_switch?.isChecked = item.isChecked
        view.filter_item_switch?.tag = position
        // separator is visible unless last element
        view.filter_item_separator?.visibility = if (position == count - 1) View.GONE else View.VISIBLE
        return view
    }

    // ----------------------------------
    // View Holder
    // ----------------------------------
    inner class GuideFilterViewHolder(v: View) {
        init {
            if (!isHelpOnly) {
                v.filter_item_switch?.setOnCheckedChangeListener { buttonView, isChecked ->
                    // if no tag, exit
                    if (buttonView.tag != null) {
                        getItem(buttonView.tag as Int).isChecked = isChecked
                    }
                }
            }
            else {
                v.filter_item_switch?.visibility = View.INVISIBLE
            }
        }
    }

    // ----------------------------------
    // Filter Item
    // ----------------------------------
    class GuideFilterItem(var categoryType: CategoryType, var isChecked: Boolean)

    // ----------------------------------
    // Implementation
    // ----------------------------------
    init {
        val guideFilter = instance
        for (i in 1 until CategoryType.values().size) {
            val categoryType = CategoryType.values()[i]
            if (categoryType != CategoryType.PARTNERS) {
                items.add(GuideFilterItem(categoryType, guideFilter.valueForCategoryId(categoryType.categoryId)))
            }
        }
    }
}